File: src/regexp_ext_ffi.erl
-module(regexp_ext_ffi).

-export([capture_names/2, split_before/2, split_after/2]).

names(Regex) ->
    {namelist, List} = re:inspect(Regex, namelist),
    List.

capture_names(Regex, Subject) ->
    Names = names(Regex),
    case re:run(Subject, Regex, [{capture, Names, list}, global]) of
        {match, Captured} ->
            lists:map(
                fun(X) ->
                    lists:map(
                        fun(S) ->
                            {Name, Value} = S,
                            {named_capture, Name, unicode:characters_to_binary(Value)}
                        end,
                        lists:zip(Names, X)
                    )
                end,
                Captured
            );
        match ->
            [[]];
        nomatch ->
            []
    end.

split_after(Regex, Subject) ->
  case re:run(Subject, Regex, [{capture, first, index}]) of
    {match, [{Start,Len}]} ->
      <<Before:(Start+Len)/binary, After/binary>> = Subject,
      {some, {unicode:characters_to_binary(Before), unicode:characters_to_binary(After)}};
    nomatch -> none
  end.

split_before(Regex, Subject) ->
  case re:run(Subject, Regex, [{capture, first, index}]) of
    {match, [{Start,_}]} ->
      <<Before:(Start)/binary, After/binary>> = Subject,
      {some, {unicode:characters_to_binary(Before), unicode:characters_to_binary(After)}};
    nomatch -> none
  end.

File: src/regexp_ext_ffi.mjs
import { List } from "./gleam.mjs";
import {Some, None} from "../gleam_stdlib/gleam/option.mjs";
import { NamedCapture as RegexNamedCapture } from "./regexp_ext/regexp_ext.mjs";

export function capture_names(regex, string, names) {
  regex.lastIndex = 0;
  const matches = Array.from(string.matchAll(regex)).map((match) => {
    if (match.groups !== undefined) {
      return List.fromArray(
        Object.entries(match.groups).map(([key, value]) => {
          return new RegexNamedCapture(key, value);
        }),
      );
    } else {
      return List.fromArray([]);
    }
  });
  return List.fromArray(matches);
}

export function split_after(regex, string) {
  regex.lastIndex = 0;
  const match = regex.exec(string);
  if (match) {
    const matchEnd = match.index + match[0].length;
    const before = string.slice(0, matchEnd);
    const after = string.slice(matchEnd);
    return new Some([before, after])
  }
  return new None()
}

export function split_before(regex, string) {
  regex.lastIndex = 0;
  const match = regex.exec(string);
  if (match) {
    const matchBegin = match.index;
    const before = string.slice(0, matchBegin);
    const after = string.slice(matchBegin);
    return new Some([before, after])
  }
  return new None()
}
